libbpf-tools/kprobe_helpers.cc

```diff
@@ -0,0 +1,178 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2021 Chenyue Zhou */
+#include <set>
+#include <algorithm>
+#include <string>
+#include <regex>
+#include <fstream>
+#include <sstream>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "kprobe_helpers.h"
+
+using namespace std;
+
+#define startswith(str, prefix) ((str).find(prefix) == 0)
+
+/**
+ * get_kprobe_functions - get matched functions from `/proc/kallsyms`.
+ * @pattern: Pattern string used for regular matching.
+ * @list:    Pointer to an array of string used to store the matched function name.
+ * @sz:      The initial/final length of the array (*list).
+ *
+ * RETURN VALUE
+ *
+ *  On success 0 is returned.
+ *
+ * ERRORS
+ *  ERANGE: The number of matched functions is large than `KPROBE_LIMIT`.
+ *          `sz` was set to `KPROBE_LIMIT`, and `KPROBE_LIMIT` matched
+ *          functions in `list`.
+ *  ENOMEM: Not enough memory for intermediate variables or final results.
+ *  OTHERR: Failed to read from `/proc/kallsyms`. `sz` was set to 0.
+ */
+int get_kprobe_functions(const char *pattern, char ***list, size_t *sz) {
+  string line;
+  set<string> kprobe_blacklist_set, matched_function_set;
+  ifstream kprobe_blacklist_stream { KPROBE_BLACKLIST_FILE };
+
+  while (getline(kprobe_blacklist_stream, line)) {
+    size_t pos = line.find("\t");
+    if (pos != string::npos) {
+      line = line.substr(pos + 1);
+      kprobe_blacklist_set.insert(line);
+    }
+  }
+
+  size_t count = 0, list_size = (*sz) < 0 ? 0 : (*sz);
+  int in_init_section = 0, in_irq_section = 0;
+  regex pt("^.*\\.cold(\\.\\d+)?$");
+  regex event_re (pattern);
+  ifstream kallsyms_stream { "/proc/kallsyms" };
```

Comment:
`trace_helpers.h` already has something like this.

Reply:
Thank you for reminding.

I just read "/proc/kallsyms" and follow my logic for each matching kernel function. The functions provided by 'trace_helpers' do not seem very useful for this.

Reply:
Why?

Reply:
'trace_helper' provides a function to get all kernel function names directly. But I want to process the kernel function name at the same time as getting each kernel function name, which is faster.